crates/block-producer/src/mempool/inflight_state/mod.rs

```diff
@@ -0,0 +1,575 @@
+            nullifiers.extend(tx.get_nullifiers());
+            output_notes.extend(tx.output_notes().iter().map(|note| note.id()));
+        }
+
+        Self {
+            account_transactions,
+            nullifiers,
+            output_notes,
+        }
+    }
+}
+
+impl InflightState {
+    /// Appends the transaction to the inflight state.
+    ///
+    /// This operation is atomic i.e. a rejected transaction has no impact of the state.
+    pub fn add_transaction(
+        &mut self,
+        tx: &ProvenTransaction,
+        input_account_hash: Option<Digest>,
+    ) -> Result<BTreeSet<TransactionId>, VerifyTxError> {
+        // Separate verification and state mutation so that a rejected transaction
+        // does not impact the state (atomicity).
+        self.verify_transaction(tx, input_account_hash)?;
+
+        let parents = self.insert_transaction(tx);
+
+        Ok(parents)
+    }
+
+    fn verify_transaction(
+        &self,
+        tx: &ProvenTransaction,
+        input_account_hash: Option<Digest>,
+    ) -> Result<(), VerifyTxError> {
+        // Ensure current account state is correct.
+        let current = self
+            .accounts
+            .get(&tx.account_id())
+            .and_then(|account_state| account_state.current_state())
+            .copied()
+            .or(input_account_hash);
+        let expected = tx.account_update().init_state_hash();
+
+        if expected != current.unwrap_or_default() {
+            return Err(VerifyTxError::IncorrectAccountInitialHash {
+                tx_initial_account_hash: expected,
+                current_account_hash: current,
+            });
+        }
+
+        // Ensure nullifiers aren't already present.
+        let double_spend = tx
+            .get_nullifiers()
+            .filter(|nullifier| self.nullifiers.contains(nullifier))
+            .collect::<Vec<_>>();
+        if !double_spend.is_empty() {
+            return Err(VerifyTxError::InputNotesAlreadyConsumed(double_spend));
+        }
+
+        // Ensure output notes aren't already present.
+        let duplicates = tx
+            .output_notes()
+            .iter()
+            .map(OutputNote::id)
+            .filter(|note| self.output_notes.contains_key(note))
+            .collect::<Vec<_>>();
+        if !duplicates.is_empty() {
+            return Err(VerifyTxError::OutputNotesAlreadyExist(duplicates));
+        }
+
+        // Ensure that all unauthenticated notes have an inflight output note to consume.
+        //
+        // We don't need to worry about double spending them since we already checked for
+        // that using the nullifiers.
+        let missing = tx
+            .get_unauthenticated_notes()
+            .map(|note| note.id())
+            .filter(|note_id| !self.output_notes.contains_key(note_id))
+            .collect::<Vec<_>>();
+        if !missing.is_empty() {
+            return Err(VerifyTxError::UnauthenticatedNotesNotFound(missing));
+        }
```

Comment:
Oh right. I think I had this check/change of state as part of the pre-mempool checks in the server handler function.

Which I (mistakenly) removed when I reverted my newtype changes to simplify the PR.

I'll add these checks here; though ideally they would be pre-mempool checks and changes. This will require something like:

```rust
pub struct Transaction {
    tx: ProvenTransaction,
    /// Notes found in the store
    found_notes: BTreeSet<NoteId>,
}
```
or even better, having the found notes be the note inclusion proofs. We can make this change now already; it would just mean some adjustments and a second store call to get `note_authenticate_info` in the handler. Best would ofcourse be to fold this store call into tx inputs.

Reply:
I think I'll punt this work to a new PR. I'll add the checks to `InflightState`, including moving the current nullifier precheck in here.

The PR to add this should be much easier to review without all this noise then.

Reply:
Actually scratch that. Creating the test data is a pain this way. I'll add back some new types which should make intent more clear.

Reply:
I've pushed a commit with some new types which make things more explicit, and correct I think. lmkwyt